### PR TITLE
Add sections on conda and Snakemake

### PR DESCRIPTION
**Pull request by f-heeger**
I wrote sections on conda and sankemake and just added them as points to the main website.

### CODE REVIEW

File: conda.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+Installation
+------------
+
+Conda can be installed as part of the [Miniconda](https://docs.conda.io/en/latest/miniconda.html) package. It works on Windows, MacOSX and Linux. See the website for further installation instructions.  
+
+Channels
+--------
+Software packages for conda (also called `recipes`) are provided in repositories called `channels`. Channels can be created by anyone and many organizations or even individual researchers maintain their own channels. By default conda will install packages from the main channel at [anaconda.com](https://repo.anaconda.com/pkgs/), but you specify other channels, when installing packages (see below). Useful channels include [`conda-forge`](https://anaconda.org/conda-forge) for community maintained packages, [`r`](https://anaconda.org/r) for R packages, and [`bioconda`](https://anaconda.org/bioconda) for bioinformatics packages. You can use the search on the [Anaconda main page](https://anaconda.org/) to search for channels, that provide a certain package. Just remember, that anyone can provide a channel and make sure that the channel is reliable.
```
Comment (Psirus) [inline conda.md]: Unfortunately the whole paragraph is one line, a bit harder to read the diff. Anyway, there are a couple of superfluous commas here:
> ... but you [can] specify other channels when installing packages ...

> ... to search for channels that provide a certain package ...

> Just remember that anyone can provide ...

For the last sentence I'd add that you should see if they are trustworthy, and not trying to run a bitcoin miner on your machine :)

File: conda.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+~~~
+
+### Deactivating (leaving) an environment
+You can deactivate an environment with the command:
+~~~sh
+conda deactivate
+~~~
+
+### Installing packages in an environment
+Once you activated an environment you can install additional packages withe `install` command. For example to install `doit` use:
```
Comment (Psirus) [inline conda.md]: > with the `install` command

File: conda.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+Managing your environments
+--------------------------
+This is only a short introduction on managing you environments. For full details see the conda documentation on this topic [here](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html).
+
+### Creating an empty environment
+~~~sh
+conda create -n exampleEnv
+~~~
+The easiest way to create an environment is to just specify its name (with `-n`). This will create an empty environment in that you can install packages late (see below).
```
Comment (Psirus) [inline conda.md]: > This will create an empty environment in **which** you can install packages **later** (see below).

File: conda.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+This is only a short introduction on managing you environments. For full details see the conda documentation on this topic [here](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html).
+
+### Creating an empty environment
+~~~sh
+conda create -n exampleEnv
+~~~
+The easiest way to create an environment is to just specify its name (with `-n`). This will create an empty environment in that you can install packages late (see below).  
+
+### Activating (entering) an environment
+To work with your environment, you have to activate it. While an environment is activated all packages that were installed into it are available as if the were installed system wide.
```
Comment (Psirus) [inline conda.md]: > While an environment is activated, all packages that were installed into it are available as if **they** were installed **system-wide**.

File: conda.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+To work with your environment, you have to activate it. While an environment is activated all packages that were installed into it are available as if the were installed system wide.  
+To activate your newly created environment you use:
+~~~sh
+conda activate exampleEnv
+~~~
+Depending on your conda installation, this might not work. In this case on Linux you can try:
+~~~sh
+source activate exampleEnv
+~~~
+After the environment is activated the command prompt of your terminal should change to include the name of the environment for example it might look like this:
```
Comment (Psirus) [inline conda.md]: > ... name of the environment. For example, it might look like this:

File: conda.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+name: exampleEnv
+channels:
+    - conda-forge
+    - bioconda
+dependencies:
+    - biopython=1.78
+    - bioconductor-mgsa=1.38.0
+~~~
+
+The environment can be create form this file with the command:
```
Comment (Psirus) [inline conda.md]: > The environment can be **created from** this file with the command:

File: snakemake.md [hunk 7]
```diff
@@ -0,0 +1,223 @@
+Snakemake
+=========
+
+
+What is it?
+-----------
+`Snakemake` is a python based workflow management system that allows for whole analysis pipelines to be automatized. It lets you define workflows as a set of rules, that state how to produce one or more output files from one or more input files. Dependencies between rules will be automatically inferred on execution and independent rules will be evaluated in parallel. If the output files for a rule are newer than the input files it will not be executed again. Snakemake can directly use conda or docker for reproducible software environments.
```
Comment (Psirus) [inline snakemake.md]: Try to make each sentence one line for better diffs. Anyway:
> It lets you define workflows as a set of rules that state how to produce one or more output files from one or more input files.

(Had superfluous comma.)

File: snakemake.md [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+`Snakemake` is a python based workflow management system that allows for whole analysis pipelines to be automatized. It lets you define workflows as a set of rules, that state how to produce one or more output files from one or more input files. Dependencies between rules will be automatically inferred on execution and independent rules will be evaluated in parallel. If the output files for a rule are newer than the input files it will not be executed again. Snakemake can directly use conda or docker for reproducible software environments.  
+
+
+Why should I use it?
+--------------------
+Snakemake allows to automatize data analysis. On the one hand this means that analysis can be quickly repeated if necessary (e.g. when new data is available). On the other hand automation drastically reduces the potential for human error. The rule based structure of Snakemake enforces a modular design, that improves organization of the code and helps with re-usability of analysis steps. Additionally the integration with conda allows for automatic installation of required software as well as the explicit selection and documentation of software versions, which is necessary for reproducibility.
```
Comment (Psirus) [inline snakemake.md]: > The **rule-based** structure of Snakemake enforces a modular design that improves organization of the code and helps with **reusability** of analysis steps.

File: snakemake.md [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+------------
+Snakemake can be installed via `conda`:
+~~~sh
+conda install -c conda-forge snakemake
+~~~
+or via `pip`
+~~~sh
+pip install snakemake
+~~~
+note that some of the more advances features (like report generation) might not be available when you install via `pip`.
```
Comment (Psirus) [inline snakemake.md]: > **Note** that some ...

File: snakemake.md [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+
+Basic use
+---------
+Snakemake has many features and options. This is only a very small introduction. You can find the full documentation [here](https://snakemake.readthedocs.io).  
+   
+A Snakemake workflow consist of a set of rules, that are implicitly linked by there input and output files. A rule normally has at least one input file, at least one output file and defines how to produce one from the other. A minimal toy example (that just copies a file) might look like this:
```
Comment (Psirus) [inline snakemake.md]: > A Snakemake workflow consist of a set of rules that are implicitly linked by **their** input and output files.

Also note the deleted comma.

File: snakemake.md [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+~~~python
+rule exampleCp1:
+    input: "inFile.txt"
+    output: "outFile.txt"
+    shell:
+        "cp {input} {output}"
+~~~
+Here the actual process of the rule is defined with the `shell` keyword. This means the following string will directly be interpreted as a command to the operating system.
```
Comment (Psirus) [inline snakemake.md]: I would say the commands are interpreted by your shell, rather than your OS.

File: snakemake.md [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+
+~~~
+  fileA.txt ──> fileB.txt ─┐
+                           │
+                           ├──> fileD.txt ──> fileE.txt
+                           │
+  fileC.txt ───────────────┘
+~~~
+
+If you write your rules into a file called `Sankefile` you can then execute them by just calling
```
Comment (Psirus) [inline snakemake.md]: > **`Snakefile`**

File: snakemake.md [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+~~~sh
+snakemake
+~~~
+in the folder where the folder where the file is. If you give your file a different name (e.g. `main.snakemake.py`), you have to use the `-s` option to execute your worklflow:
```
Comment (Psirus) [inline snakemake.md]: > in the folder where the file is.

File: snakemake.md [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+~~~sh
+sankemake -s main.snakemake.py
+~~~
+
+Wildcards
+---------
+In many cases you want a rule executed for multiple files (e.g. different samples). you can use wildcards in input and output file names to achieve this. For example:
```
Comment (Psirus) [inline snakemake.md]: > ... for multiple files (e.g. different samples). **You** can use wildcards ...

File: snakemake.md [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+    shell:
+        "grep \"YES\" {input} > {output}"
+~~~
+
+If you want to run this for sample 1-4 you can write your `all` rule like this:
+~~~python
+rule all:
+    input: expand("sample{sNr}_filtered.txt", sNr=[1,2,3,4])
+~~~
+You can also use multiple wildcards in one file name and use wildcards as values in you command, for example to try out different values for parameters for the filtering in the example above:
```
Comment (Psirus) [inline snakemake.md]: > use wildcards as values in **your** command, ...

File: snakemake.md [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+    input: "data.tsv"
+    output: "clusters.txt"
+    shell:
+        "kMeansCluster -k {config[k]} -d {config[dist]} -o {output} {input}"
+~~~
+
+Conda integration
+-----------------
+Snakemake offers direct integration with [conda](docs.conda.io), which allows to fix software versions as well as  automation of installation.
+For this you have to create a conda environment file, that defines the software, the version you want and the conda `channel` it can be found in (if not the default). For example for [Blast](https://blast.ncbi.nlm.nih.gov/Blast.cgi) this environment file (`blast.yaml`) would look like this:
```
Comment (Psirus) [inline snakemake.md]: > ... create a conda environment file that defines the software, the version ...

Superfluous comma after environment file.

File: snakemake.md [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+    conda:
+        "blast.yaml"
+    shell:
+        "blastn -query {input.query} -db {input.sb} -out {output}
+~~~
+To run this using conda you have to use the `--use-conda` option when you execute sankemake:
+~~~sh
+snakemake --use-conda
+~~~
+Sankemake will than use conda (which has to be installed) to automatically install Blast and run it.
```
Comment (Psirus) [inline snakemake.md]: > **Snakemake** will **then** use conda ...

**Review by Psirus — APPROVED** [review]
Great stuff! Thank you very much.

First, you should add yourself to the authors list (in `conf.py`). Only if you want to, of course.

Secondly, now that we have two buildtools, the "why" section of doit and snakemake read pretty much the same. Maybe we should have a comparison somewhere? At the very least snakemake should be added to the "automate everything" section of the introduction (`reproducibility.md`) alongside pydoit.

I am very much in favour of merging this. I've learned a lot reading through it, so again, thank you. The rest are some typos I've discovered along the way. Feel free to ignore them :)

File: conda.md (hunk 1 above)
Comment (f-heeger) [reply conda.md]: I think some of the commas are not strictly necessary, but improve readability. So unless you think they are strictly wrong, I would like to keep them.

**Comment by TTitscher** [comment]
To combine the stable conda environment with my (currently) favorite tool pydoit, I have to run pydoit _inside_ a conda environment. This means that the environment must include _all_ dependencies of the whole workflow, e.g. "run a huge simulation", "do post processing", "run latex",  ...  This is not modular and conda itself has a hard time resolving potential versioning conflicts. 
It seems like with snakemake, we can run conda _within_ snakemake, which solves this issue. I will definitely try that out. Thank you for this input!

> Secondly, now that we have two buildtools, the "why" section of doit and snakemake read pretty much the same. 

I fully agree and with more and more content, we should also think about a more structured TOC, e.g. with _workflow management tools_ section including both snakemake and pydoit and potentially others. That would IMO be another PR.

File: snakemake.md [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+~~~
+To run this using conda you have to use the `--use-conda` option when you execute sankemake:
+~~~sh
+snakemake --use-conda
+~~~
+Snakemake will then use conda (which has to be installed) to automatically install Blast and run it.
+
+Other features
+--------------
+Snakemake is a powerful tool, that offers many additional features. You can find more details in the [documentation](snakemake.readthedocs.io). Interesting features include:
```
Comment (CagtayFabry) [inline snakemake.md]:
```suggestion
Snakemake is a powerful tool, that offers many additional features. You can find more details in the [documentation](https://snakemake.readthedocs.io). Interesting features include:
```
thanks for the write-up @f-heeger 👍

**Pull request merged**